### PR TITLE
PR: User profile controller

### PR DESCRIPTION
**Pull request by TymofiiSavinov**
Store user as user profile. Created methods:

1.  set_supply_amount, 
2. set_borrow_amount,
3. is_exist,
4. get_profile_index.

### CODE REVIEW

**Review by pahomas — CHANGES_REQUESTED** [review]

File: contracts/dtoken/src/withdraw.rs [hunk 1]
```diff
@@ -19,20 +20,25 @@ impl Contract {
 
 #[near_bindgen]
 impl Contract {
-
     pub fn withdraw(&mut self, dtoken_amount: WBalance) -> Promise {
-        return underlying_token::ft_balance_of(
+        let user_profile = UserProfileDtoken::default();
```
Comment (markblaize) [inline contracts/dtoken/src/withdraw.rs:24]: @TymofiiSavinov UserProfile should be part of Dtoken storage

File: contracts/dtoken/src/user_profile.rs [hunk 2]
```diff
@@ -0,0 +1,130 @@
+use near_sdk::collections::LookupMap;
+use crate::*;
+
+#[derive(BorshSerialize)]
+pub enum Action {
+    Supply,
+    Borrow,
+}
+
+#[near_bindgen]
```
Comment (markblaize) [inline contracts/dtoken/src/user_profile.rs:10]: @TymofiiSavinov we don't need a bindgen for internal structures

File: contracts/dtoken/src/user_profile.rs [hunk 3]
```diff
@@ -0,0 +1,130 @@
+use near_sdk::collections::LookupMap;
+use crate::*;
+
+#[derive(BorshSerialize)]
+pub enum Action {
```
Comment (markblaize) [inline contracts/dtoken/src/user_profile.rs:5]: @TymofiiSavinov do we use this anywhere?

File: contracts/dtoken/src/user_profile.rs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    ///  Supply user amount
+    pub(crate) supply_amount: WBalance,
+
+    ///  Borrow user amount
+    pub(crate) borrow_amount: WBalance,
+}
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct UserProfileDtoken {
```
Comment (markblaize) [inline contracts/dtoken/src/user_profile.rs:22]: @TymofiiSavinov we don't need a separate wrapper for user profiles, just make `user_profile_list` a part of dtoken

File: contracts/dtoken/src/supply.rs [hunk 5]
```diff
@@ -4,23 +4,30 @@ use crate::*;
 impl Contract {
     #[payable]
     pub fn supply(&mut self, token_amount: WBalance) -> PromiseOrValue<U128> {
+        let user_profile = UserProfileDtoken::default();
+        let supply_amount = user_profile.get_profile(self.get_signer_address().clone()).supply_amount as u128;
+        if user_profile.is_exist(self.get_signer_address().clone()) &&
```
Comment (markblaize) [inline contracts/dtoken/src/supply.rs:9]: @TymofiiSavinov 
we should create profile on Supply (if doesn't exist) 
and we should prevent Withdraw, Borrow, Repay if user profile doens't exist

**Review by markblaize — CHANGES_REQUESTED** [review]

File: contracts/controller/src/user_profile.rs [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+impl Default for UserProfileController {
+    fn default() -> Self {
+        Self {
+            user_profiles: LookupMap::new(b"s".to_vec()),
+        }
+    }
+}
+
+#[near_bindgen]
+impl UserProfileController {
```
Comment (markblaize) [inline contracts/controller/src/user_profile.rs:26]: @TymofiiSavinov This should be part of Controller not a separate contract

**Pull request closed without merging**